rfcs/3-ksn-api.md

```diff
@@ -0,0 +1,112 @@
+    // This callback will be called directly on KSN.
+    // It is impossible to perform writes on this callback
+    // the callback can invoke a background task to write data but it will not be atomic.
+    client.run_on_background(async function(async_client) {
+        // here it is possible to perform writes (after entering an atomic block).
+        async_client.block((client)=>{
+                return client.call('set', 'x', '1');
+        });
+    });
+});
+```
+
+## Option 2 - Post notification job only
+
+In this option we will spare the JS user from all the details on KSN and post jobs. We will accumulate all the events for the user and call it on a post job where it is safe to write. The main disadvantage it that the JS user can not look at the data at the time of the notification. He can only read the data at the end, so in case of deleted key it will not be possible to see the deleted value. The advantage is API simplicity.
+
+Example:
+
+```js
+redis.register_notifications_consumer("consumer", "", function(client, data){
+    // this callback will be called after all the KSN finished.
+    // it is safe to write and read.
+    // the callback has access to the notification name and key name.
+    // The callback has not access to the data at the time the notification happened.
+    let key_name = data.key;
+    client.call('set', 'x', '1'); // performing write command
+});
+```
+
+## Option 3 - KSN callback + Post Notification job
+
+This option is a combination on 2 and 1. The callback will be called on the KSN event without the ability to write. We will expose an API to register a post notification job in which is is safe to write.
+
+Example:
+
+```js
+redis.register_notifications_consumer("consumer", "", function(client, data){
+    // This callback will be called directly on KSN.
+    // It is impossible to perform writes on this callback
+    // the callback can register a post notification job
+    // to write data atomically with the notification.
+    client.register_post_notification_job(function(client) {
+        // here we can write
+        client.call('set', 'x', '1');
+    });
+});
+```
+
+## Option 4 - Simple API for regular users + advance API for advance users
+
+This option is a combination on 2 and 4. We will implement a simple API for most users which will look like option 2. For advance users we will implement option 4.
+
+Example:
+
+```js
+// Simple API for most users
+redis.register_notifications_consumer("consumer", "", function(client, data){
+    // this callback will be called after all the KSN finished.
+    // it is safe to write and read.
+    // the callback has access to the notification name and key name.
+    // The callback has not access to the data at the time the notification happened.
+    let key_name = data.key;
+    client.call('set', 'x', '1'); // performing write command
+});
+
+// Advance API for advance users
+redis.register_ksn_consumer("consumer", "", function(client, data){
```

Comment:
can we make the `ksn` in the function name an optional parameter of the function, so it's the same API?

Reply:
Its a key space notification indicating key space changed. Just trigger does give you any context of why it has been triggered.

Reply:
register_crud_trigger?